compiler/rustc_builtin_macros/src/assert.rs

```diff
@@ -154,10 +161,66 @@ fn parse_assert<'a>(cx: &mut ExtCtxt<'a>, sp: Span, stream: TokenStream) -> PRes
         return parser.unexpected();
     }
 
-    Ok(Assert { cond_expr, custom_message })
+    let cond_expr = expand_cond(cx, parser, inner_cond_expr.clone());
+    Ok(Assert { cond_expr, inner_cond_expr, custom_message })
+}
+
+fn expand_cond(cx: &ExtCtxt<'_>, parser: Parser<'_>, cond_expr: P<Expr>) -> P<Expr> {
+    let span = cx.with_call_site_ctxt(cond_expr.span);
+    // Coerce the expression to `bool` for more accurate errors. If `assert!` is passed an
+    // expression that isn't `bool`, the type error will point at only the expression and not the
+    // entire macro call. If a non-`bool` is passed that doesn't implement `trait Not`, we won't
+    // talk about traits, we'll just state the appropriate type error.
+    // `let assert_macro: bool = $expr;`
+    let ident = Ident::new(sym::assert_macro, DUMMY_SP);
+
+    let expr = if use_assert_2024(span) {
+        // `{ let assert_macro: bool = $expr; assert_macro }`
+        cond_expr
+    } else {
+        // In <=2021, we allow anything that can be negated to `bool`, not just `bool`s. We use the
+        // "double not" trick to coerce the expression to `bool`. We still assign it to a new `bool`
+        // binding so that in the case of a type that implements `Not` but doesn't return `bool`,
+        // like `i32`, we still point at the condition and not at the whole macro.
+        // `{ let assert_macro: bool = !!$expr; assert_macro }`
+        let not = |expr| parser.mk_expr(span, ast::ExprKind::Unary(ast::UnOp::Not, expr));
+        not(not(cond_expr))
+    };
+    let block = thin_vec![
+        cx.stmt_let_ty(
+            DUMMY_SP,
+            false,
+            ident,
+            Some(cx.ty_ident(span, Ident::new(sym::bool, DUMMY_SP))),
```

Comment:
The hygiene is not sufficient here. This example now breaks:

```rust
fn main() {
    type bool = i32;
    assert!(true);
}
```

```
error[E0308]: mismatched types
 --> src/main.rs:3:13
  |
3 |     assert!(true);
  |             ^^^^ expected `i32`, found `bool`
  |
```

Reply:
It's probably better and easier to just use `if <expr> {} else { panic!(..) }`, without making any block expressions or local bindings, and without making any changes to the expansion in previous editions.